main.go

```diff
@@ -317,25 +283,93 @@ func main() {
 	}
 }
 
-func before(c *cli.Context) (*state.HelmState, helmexec.Interface, error) {
-	file := c.GlobalString("file")
+func eachDesiredStateDo(c *cli.Context, converge func(*state.HelmState, helmexec.Interface) []error) error {
+	fileOrDirPath := c.GlobalString("file")
+	desiredStateFiles, err := findDesiredStateFiles(fileOrDirPath)
+	if err != nil {
+		return err
+	}
+	for _, f := range desiredStateFiles {
+		state, helm, err := loadDesiredStateFromFile(c, f)
+		if err != nil {
+			return err
+		}
+		errs := converge(state, helm)
+		if err := clean(state, errs); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func findDesiredStateFiles(specifiedPath string) ([]string, error) {
+	var existingDir string
+	if specifiedPath != "" {
+		if !pathExists(specifiedPath) {
+			return []string{}, fmt.Errorf("specified state file %s is not found", specifiedPath)
+		}
+
+		if !directoryExists(specifiedPath) {
+			return []string{specifiedPath}, nil
+		}
+
+		existingDir = specifiedPath
+	} else if pathExists(DefaultHelmfileDirectory) && directoryExists(DefaultHelmfileDirectory) {
+		existingDir = DefaultHelmfileDirectory
+	}
+
+	var existingFile string
+	if pathExists(DefaultHelmfile) {
+		existingFile = DefaultHelmfile
+	} else if pathExists(DeprecatedHelmfile) {
+		log.Printf(
+			"warn: %s is being loaded: %s is deprecated in favor of %s. See https://github.com/roboll/helmfile/issues/25 for more information",
+			DeprecatedHelmfile,
+			DeprecatedHelmfile,
+			DefaultHelmfile,
+		)
+		existingFile = DeprecatedHelmfile
+	}
+
+	if specifiedPath == "" && existingFile != "" && existingDir != "" {
```

Comment:
I believe that `specifiedPath == ""` is needed in order to allow the user to selectively run `helmfile sync` for the default helmfile.d or helmfile.yaml, and `helmfile sync -f $path` for a specific helmfile

Reply:
The logic looks correct after stepping through a bunch of options, but it's very complicated.  I think I would move the DefaultHelmfile/DeprecatedHelmfile check as part of the first if else.  I also think you can drop the `specifiedPath==""` from this if statement as the first function should always set existingDir if specifiedPath is set or return.

And of course I could be missing something on my recommendations.

Reply:
Ah, so this block should be nested inside the `} else if pathExists(DefaultHelmfileDirectory) && directoryExists(DefaultHelmfileDirectory) {` block in the first if-else?
Sounds good! I'll address it asap. Thanks 🙏